bftengine/src/bftengine/FullNodeReplica.cpp

```diff
@@ -0,0 +1,295 @@
+#include "util/json_output.hpp"
+#include "SharedTypes.hpp"
+#include "communication/StateControl.hpp"
+
+using concordUtil::Timers;
+
+namespace bftEngine::impl {
+
+FullNodeReplica::FullNodeReplica(const ReplicaConfig &config,
+                                 std::shared_ptr<IRequestsHandler> requestsHandler,
+                                 IStateTransfer *stateTransfer,
+                                 std::shared_ptr<MsgsCommunicator> msgComm,
+                                 std::shared_ptr<MsgHandlersRegistrator> msgHandlerReg,
+                                 concordUtil::Timers &timers,
+                                 MetadataStorage *metadataStorage)
+    : ReplicaForStateTransfer(config, requestsHandler, stateTransfer, msgComm, msgHandlerReg, true, timers),
+      fn_metrics_{metrics_.RegisterCounter("receivedCheckpointMsgs"),
+                  metrics_.RegisterCounter("sentAskForCheckpointMsgs"),
+                  metrics_.RegisterCounter("receivedInvalidMsgs"),
+                  metrics_.RegisterGauge("lastExecutedSeqNum", lastExecutedSeqNum)},
+      metadataStorage_{metadataStorage} {
+  LOG_INFO(GL, "Initialising Full Node Replica");
+  repsInfo = new ReplicasInfo(config, dynamicCollectorForPartialProofs, dynamicCollectorForExecutionProofs);
+  msgHandlers_->registerMsgHandler(
+      MsgCode::Checkpoint, std::bind(&FullNodeReplica::messageHandler<CheckpointMsg>, this, std::placeholders::_1));
+  msgHandlers_->registerMsgHandler(
+      MsgCode::ClientRequest,
+      std::bind(&FullNodeReplica::messageHandler<ClientRequestMsg>, this, std::placeholders::_1));
+  msgHandlers_->registerMsgHandler(
+      MsgCode::StateTransfer,
+      std::bind(&FullNodeReplica::messageHandler<StateTransferMsg>, this, std::placeholders::_1));
+  metrics_.Register();
+
+  SigManager::init(config_.replicaId,
+                   config_.replicaPrivateKey,
+                   config_.publicKeysOfReplicas,
+                   concord::crypto::KeyFormat::HexaDecimalStrippedFormat,
+                   ReplicaConfig::instance().getPublicKeysOfClients(),
+                   concord::crypto::KeyFormat::PemFormat,
+                   {{repsInfo->getIdOfOperator(),
+                     ReplicaConfig::instance().getOperatorPublicKey(),
+                     concord::crypto::KeyFormat::PemFormat}},
+                   *repsInfo);
+
+  // Register status handler for Full Node replica
+  registerStatusHandlers();
+  bft::communication::StateControl::instance().setGetPeerPubKeyMethod(
+      [&](uint32_t id) { return SigManager::instance()->getPublicKeyOfVerifier(id); });
+}
+
+void FullNodeReplica::start() {
+  ReplicaForStateTransfer::start();
+  size_t sendAskForCheckpointMsgPeriodSec = config_.get("concord.bft.ro.sendAskForCheckpointMsgPeriodSec", 30);
+  askForCheckpointMsgTimer_ = timers_.add(
+      std::chrono::seconds(sendAskForCheckpointMsgPeriodSec), Timers::Timer::RECURRING, [this](Timers::Handle) {
+        if (!this->isCollectingState()) {
```

Comment:
I think that you should state in the commit and also very clearly here, that this is temporary code that should be removed later.
FN should not send AskForCheckpointMsg on startup, this is RO replica behavior.It might do something similar from another module that has not yet been merged.
This comment is valid for all places where RO replica behavior is temporarily adapted.

Reply:
will add a comment  in the beginning of the file and in the PR description i will mention about the same